rust/src/dataset/transaction.rs

```diff
@@ -0,0 +1,555 @@
+                // Rewrite is only compatible with operations that don't touch
+                // existing fragments.
+                // TODO: it could also be compatible with operations that update
+                // fragments we don't touch.
+                Operation::Append { .. } => false,
+                Operation::Delete {
+                    updated_fragments,
+                    deleted_fragment_ids,
+                    ..
+                } => {
+                    // If we rewrote any fragments that were modified by delete,
+                    // we conflict.
+                    let left_ids: HashSet<u64> = old_fragments.iter().map(|f| f.id).collect();
+                    deleted_fragment_ids.iter().any(|f| left_ids.contains(f))
+                        || updated_fragments.iter().any(|f| left_ids.contains(&f.id))
+                }
+                Operation::Rewrite {
+                    old_fragments: right_fragments,
+                    ..
+                } => {
+                    // As long as they rewrite disjoint fragments they shouldn't conflict.
+                    let left_ids: HashSet<u64> = old_fragments.iter().map(|f| f.id).collect();
+                    right_fragments.iter().any(|f| left_ids.contains(&f.id))
+                }
+                _ => true,
+            },
+            // Overwrite always succeeds
+            Operation::Overwrite { .. } => false,
+            Operation::CreateIndex { .. } => match &other.operation {
+                Operation::Append { .. } => false,
+                // Indices are identified by UUIDs, so they shouldn't conflict.
+                Operation::CreateIndex { .. } => false,
+                // Although some of the rows we indexed may have been deleted,
+                // row ids are still valid, so we allow this optimistically.
+                Operation::Delete { .. } => false,
+                // Merge doesn't change row ids, so this should be fine.
+                Operation::Merge { .. } => false,
+                // Rewrite likely changed many of the row ids, so our index is
+                // likely useless. It should be rebuilt.
+                // TODO: we could be smarter here and only invalidate the index
+                // if the rewrite changed more than X% of row ids.
+                Operation::Rewrite { .. } => true,
+                _ => true,
+            },
+            Operation::Delete {
+                updated_fragments: left_fragments,
+                ..
+            } => match &other.operation {
+                Operation::CreateIndex { .. } => false,
+                Operation::Delete {
+                    updated_fragments: right_fragments,
+                    ..
+                } => {
+                    // If we update the same fragments, we conflict.
+                    let left_ids: HashSet<u64> = left_fragments.iter().map(|f| f.id).collect();
+                    right_fragments.iter().any(|f| left_ids.contains(&f.id))
```

Comment:
Minor nit: This "fragments intersect" logic is reused a lot.  Might be worth pulling into a helper function for readability.